### ISSUE TITLE
IP pool quotas and utilization

### ISSUE THREAD

**Issue by augustuswm** [oxidecomputer/omicron#4958]
Discussed with @zephraph this morning about the controls we have around IP pools. I don't think these are high priority, but wanted to get these notes down.

As an operator it would be useful to see and control the utilization of the IP pools within a rack. IP pools can be assigned to zero or more silos and a silo can have zero or more ip pools.

Currently we do not provide visibility in to how utilized those resources (IP addresses) are. Both at a silo and fleet level view it would be useful to see the used and remaining number of ip addresses per pool. (Though we may want to be more restrictive about what the silo can see). Further in the future it would be even better to see that data broken by their usage type (external ip, floating ip, nat, ?).

Once we can see this data, it would also be very helpful to be able to control the use of an IP pool. Where currently we can divide up a block of IPs into multiple pools of fixed sizes and apply them to individual silos, it can be time consuming to manually manage them. In the case that I have a single large pool of IPs and would like to allocate a fixed number to a given silo it would be helpful to be able to set a quota on the (silo, pool) pair and track it with the other quotas we place on silos.

**Comment by zephraph** [comment]
This came up in a customer conversation as something that they wanted to see so it's worth fleshing this out in more detail.

Related is https://github.com/oxidecomputer/omicron/issues/5031. 

In general there's a few aspects that we want to consider here. The most important problem to tackle in the short term is one of limits: there exists the possibility when two or more silos are pulling from the same IP pool that one silo can greedily use up all the IPs and prevent other silos from instantiating any instances. 

As @augustuswm mentioned here, we want to be able to divide up an IP pool amongst different silos. The first pass thinking I had on this (and likely the easiest to implement) would be to have a rollup count of utilized IPs in the pool (as @david-crespo [suggests](https://github.com/oxidecomputer/omicron/issues/5031) and be able to set a limit per silo of a number of IPs they can consume from the pool. 

I've talked to both @augustuswm and @internet-diglett about this approach and there is some weirdness. Likely network operators would think more about blocks of addresses than top level counts like I'm suggesting. My current thinking is that associating ranges inside a pool to a specific IP is too logically complicated though. 

Would welcome other folks' thoughts on it.

**Comment by internet-diglett** [comment]
> As @augustuswm mentioned here, we want to be able to divide up an IP pool amongst different silos. The first pass thinking I had on this (and likely the easiest to implement) would be to have a rollup count of utilized IPs in the pool (as @david-crespo https://github.com/oxidecomputer/omicron/issues/5031 and be able to set a limit per silo of a number of IPs they can consume from the pool.

From an operator perspective I think this is a great first step.

> Likely network operators would think more about blocks of addresses than top level counts like I'm suggesting.

After thinking about it for a bit, in my experience I think it's not unusual to have multiple address blocks assigned to a rack or set of racks within a DC. While I suspect that some operators will want have a specific silo use a specific block of addresses, they can accomplish this by using a dedicated pool and range for that silo. If they are okay with a silo (or silos) using multiple blocks of addresses, they can assign multiple ranges to a pool. So the idea of doing rollup counts at the pool level still allows them to observe addresses at a per-block level (single pool with single range) or multi-block level (single pool with multiple ranges).

**Comment by rmustacc** [comment]
Yeah, I think I would start with having this be count based versus block based as the whole point of IP pools as you point out @internet-diglett is to consolidate disparate blocks into a single logical purpose that someone cares about and if you want to only assign a single block to someone, then you should make that its own IP pool. Later, someone may ask what is the utilization of each of the ranges I have in my pool, but I think that's a secondary planning concern versus tenant utilization.

Put differently, the idea is that you have put a whole bunch of different blocks into a pool and therefore don't care about the individual block allocation any more in theory as folks can get things from whatever block they care about. Forcing you to think about how each range is being used for usage seems a bit contrary and depends a bit on where the dividing line on management is personally.
